src/vs/editor/contrib/referenceSearch/common/referenceSearch.ts

```diff
@@ -39,4 +39,30 @@ export function provideReferences(model: IReadOnlyModel, position: Position, pro
 	});
 }
 
+export function provideExternalReferences(model: IReadOnlyModel, query: ISymbolDescriptor, hints: { [hint: string]: any }, progress: (references: IReferenceInformation[]) => void): TPromise<IReferenceInformation[]> {
+
+	// collect references from all providers
+	const promises = ExternalReferenceProviderRegistry.ordered(model).map(provider => {
+		return asWinJsPromise((token) => {
+			return provider.provideExternalReferences(query, hints, token, progress);
+		}).then(result => {
+			if (Array.isArray(result)) {
```

Comment:
Ah, I think this code exists because `provider.provideExternalReferences(query, hints, token, progress)` gets serialized to the extension host and the response is serialized back. We expect certain types here, but ultimately the main thread receives JSON from a potentially untrusted extension.